erts/emulator/beam/erl_bif_binary.c

```diff
@@ -875,22 +891,44 @@ static BFReturn bm_find_all_non_overlapping(BinaryFindContext *ctx, byte *haysta
     Sint *gs = bmd->goodshift;
     Sint *bs = bmd->badshift;
     byte *needle = bmd->x;
-    Sint i;
+    Sint i = -1; /* Use memchr on start and on every match */
     Sint j = state->pos;
     Uint m = state->m;
     Uint allocated = state->allocated;
     FindallData *out = state->out;
     register Uint reds = *reductions;
+    byte *pos_pointer;
+    Sint needle_last = blen - 1;
+    Sint mem_read;
 
-    while (j <= len - blen) {
-	if (--reds == 0) {
+    for(;;) {
+	if (i < 0) {
+	    mem_read = MIN(len - j, reds);
+	    ASSERT(mem_read > 0);
+	    pos_pointer = memchr(&haystack[j], needle[needle_last], mem_read);
+	    if (pos_pointer == NULL) {
+		j += mem_read;
+		reds -= mem_read;
+	    } else {
+		j = pos_pointer - haystack - needle_last;
+		reds -= (pos_pointer - haystack - needle_last - j);
+	    }
+	}
+	if (j > len - blen) {
+	    state->m = m;
+	    state->out = out;
+	    *reductions = reds;
+	    return (m == 0) ? BF_NOT_FOUND : BF_OK;
+	}
+	if (reds == 0) {
```

Comment:
Before we did `if(--reds == 0)` but since `memchr` may set reductions to 0, we need to check for an explicit zero before decrementing. This means we may return 0 as the number of reductions when this procedure exits! If this is not desired or not allowed, then we need to address it.